src/Functions/GatherUtils/Algorithms.h

```diff
@@ -418,15 +433,92 @@ bool sliceHasImpl(const FirstSliceType & first, const SecondSliceType & second,
                 has = true;
         }
 
-        if (has && !all)
+        if (has && search_type == ArraySearchType::Any)
             return true;
 
-        if (!has && all)
+        if (!has && search_type == ArraySearchType::All)
             return false;
+    }
+    return search_type == ArraySearchType::All;
+}
+/// Methods to check if first array has elements from second array, overloaded for various combinations of types.
+template < typename FirstSliceType,
+           typename SecondSliceType,
+           bool (*isEqual)(const FirstSliceType &, const SecondSliceType &, size_t, size_t),
+           bool (*isEqualUnary)(const SecondSliceType &, size_t, size_t)>
+bool sliceHasImplSubSeq(const FirstSliceType & first, const SecondSliceType & second, const UInt8 * first_null_map, const UInt8 * second_null_map) noexcept
+{
+    const bool has_first_null_map = first_null_map != nullptr;
+    const bool has_second_null_map = second_null_map != nullptr;
+
+    if (second.size == 0) 
+        return true;
+
+    const auto aux = buildAux<SecondSliceType, isEqualUnary>(second, second_null_map);
 
+    size_t secondCur = 0;
+    size_t firstCur = 0;
+
+    while (firstCur < first.size && secondCur < second.size)
+    {
+        auto cond_values_match = isEqual(first, second, firstCur, secondCur);
+
+        const bool is_first_null = has_first_null_map && first_null_map[firstCur];
+        const bool is_second_null = has_second_null_map && second_null_map[secondCur];
+
+        auto cond_null_match = is_first_null && is_second_null;
+        if (cond_values_match || cond_null_match)
```

Comment:
I'm afriad this condition might not work as expected when one of the two elements is null. You should use something similair to `sliceHasImplAnyAll(): line 432`

Reply:
I changed the predicate to avoid calling `isEqual` unnecessarily.